File: DesktopApp/src/main/transfer/TransferClient.ts
/**
 * TransferClient.ts
 * TCP client that connects to a peer's TransferServer and streams a file.
 * Emits typed events for progress and completion.
 * No Electron/IPC knowledge — pure Node.js networking.
 */

import { connect, Socket } from 'net';
import { EventEmitter } from 'events';
import { FileChunker } from './FileChunker';
import { randomUUID } from 'crypto';
import { TransferCrypto } from '../crypto/TransferCrypto';

export interface SendFileOptions {
  filePath: string;
  fileName: string;
  fileSize: number;
  checksum: string;
  peerIp: string;
  peerPort: number;
  senderDeviceId: string;
  senderName: string;
  resumeOffset?: number;
  /** Whether the peer advertised AES-256-GCM encryption support */
  peerSupportsEncryption?: boolean;
}

export interface TransferClientEvents {
  progress: (transferId: string, bytesSent: number, totalBytes: number, speed: number, eta: number, progress: number) => void;
  complete: (transferId: string) => void;
  error: (transferId: string, err: Error) => void;
  cancelled: (transferId: string) => void;
  declined: (transferId: string) => void;
}

export declare interface TransferClient {
  on<K extends keyof TransferClientEvents>(event: K, listener: TransferClientEvents[K]): this;
  emit<K extends keyof TransferClientEvents>(event: K, ...args: Parameters<TransferClientEvents[K]>): boolean;
}

interface ActiveSend {
  socket: Socket;
  cancel: () => void;
  /** AES-256-GCM key — set after handshake if encryption was negotiated */
  encryptionKey?: Buffer;
}

export class TransferClient extends EventEmitter {
  private activeSends: Map<string, ActiveSend> = new Map();

  /**
   * Initiate a file transfer to a peer.
   * Returns the transferId immediately; events fire asynchronously.
   */
  sendFile(options: SendFileOptions): string {
    const transferId = randomUUID();
    this.doSend(transferId, options);
    return transferId;
  }

  /**
   * Same as sendFile but uses a caller-supplied transferId.
   * Used by TransferQueue so it can register the transfer before
   * the TCP connection starts, preventing lost progress events.
   */
  sendFileWithId(transferId: string, options: SendFileOptions): void {
    this.doSend(transferId, options);
  }

  cancelTransfer(transferId: string): void {
    const send = this.activeSends.get(transferId);
    if (send) {
      send.cancel();
      this.activeSends.delete(transferId);
      this.emit('cancelled', transferId);
    }
  }

  private doSend(transferId: string, options: SendFileOptions): void {
    const {
      filePath, fileName, fileSize, checksum,
      peerIp, peerPort, senderDeviceId, senderName,
      resumeOffset = 0,
      peerSupportsEncryption = false,
    } = options;

    const socket = connect({ host: peerIp, port: peerPort }, () => {
      // Disable Nagle's algorithm so small writes (header, feedback ACKs) are
      // sent immediately without waiting to coalesce into a larger segment.
      socket.setNoDelay(true);

      // ── Build JSON header ──────────────────────────────────────────────────
      let senderKeyPair: ReturnType<typeof TransferCrypto.generateKeyPair> | null = null;

      const headerObj: Record<string, unknown> = {
        transferId,
        senderDeviceId,
        senderName,
        fileName,
        fileSize,
        checksum,
        ...(resumeOffset > 0 && { resumeOffset }),
      };

      if (peerSupportsEncryption) {
        senderKeyPair = TransferCrypto.generateKeyPair();
        headerObj['encryption'] = {
          supported: true,
          senderPublicKey: senderKeyPair.publicKeyDer.toString('base64'),
        };
      }

      socket.write(JSON.stringify(headerObj) + '\n');

      // ── Wait for accept/decline response ───────────────────────────────────
      let responseBuffer = '';

      const onResponseData = (chunk: Buffer): void => {
        responseBuffer += chunk.toString('utf8');
        const newlineIdx = responseBuffer.indexOf('\n');
        if (newlineIdx === -1) return;

        socket.removeListener('data', onResponseData);

        try {
          const response = JSON.parse(responseBuffer.slice(0, newlineIdx)) as {
            accepted: boolean;
            resumeOffset?: number;
            encryption?: { accepted: boolean; receiverPublicKey: string };
          };

          if (!response.accepted) {
            socket.destroy();
            this.activeSends.delete(transferId);
            this.emit('declined', transferId);
            return;
          }

          const serverOffset = response.resumeOffset ?? resumeOffset;

          // ── Negotiate encryption ─────────────────────────────────────────
          let encryptionKey: Buffer | undefined;

          if (
            peerSupportsEncryption &&
            senderKeyPair !== null &&
            response.encryption?.accepted === true &&
            typeof response.encryption.receiverPublicKey === 'string'
          ) {
            const receiverPubKeyDer = Buffer.from(response.encryption.receiverPublicKey, 'base64');
            encryptionKey = TransferCrypto.deriveKey(
              senderKeyPair.privateKey,
              receiverPubKeyDer,
              senderKeyPair.publicKeyDer,
              receiverPubKeyDer,
            );

            // Store key on the active send entry
            const existing = this.activeSends.get(transferId);
            if (existing) {
              this.activeSends.set(transferId, { ...existing, encryptionKey });
            }
          }

          this.streamFile(socket, transferId, filePath, fileSize, serverOffset, encryptionKey);
        } catch (err) {
          socket.destroy(err instanceof Error ? err : new Error(String(err)));
          this.emit('error', transferId, err instanceof Error ? err : new Error(String(err)));
        }
      };

      socket.on('data', onResponseData);
    });

    let cancelled = false;

    this.activeSends.set(transferId, {
      socket,
      cancel: () => {
        cancelled = true;
        socket.destroy();
      },
    });

    socket.on('error', (err) => {
      if (!cancelled) {
        this.emit('error', transferId, err);
      }
      this.activeSends.delete(transferId);
    });

    socket.on('close', () => {
      this.activeSends.delete(transferId);
    });
  }

  private streamFile(
    socket: Socket,
    transferId: string,
    filePath: string,
    fileSize: number,
    resumeOffset = 0,
    encryptionKey?: Buffer,
  ): void {
    if (encryptionKey !== undefined) {
      this.streamFileEncrypted(socket, transferId, filePath, fileSize, resumeOffset, encryptionKey);
    } else {
      this.streamFilePlaintext(socket, transferId, filePath, fileSize, resumeOffset);
    }
  }

  /** Plaintext streaming — original behaviour, unchanged. */
  private streamFilePlaintext(
    socket: Socket,
    transferId: string,
    filePath: string,
    fileSize: number,
    resumeOffset = 0,
  ): void {
    const readStream = FileChunker.createReadStream(filePath, resumeOffset > 0 ? resumeOffset : undefined);
    let cancelled = false;
    let streamDone = false;

    const existing = this.activeSends.get(transferId);
    if (existing) {
      this.activeSends.set(transferId, {
        ...existing,
        cancel: () => {
          cancelled = true;
          readStream.destroy();
          socket.destroy();
        },
      });
    }

    let feedbackBuffer = '';
    socket.on('data', (chunk: Buffer) => {
      feedbackBuffer += chunk.toString('utf8');
      let newlineIdx: number;
      while ((newlineIdx = feedbackBuffer.indexOf('\n')) !== -1) {
        const line = feedbackBuffer.slice(0, newlineIdx);
        feedbackBuffer = feedbackBuffer.slice(newlineIdx + 1);
        try {
          const fb = JSON.parse(line) as { p: number; b: number; s: number; e: number };
          if (typeof fb.p === 'number') {
            this.emit('progress', transferId, fb.b, fileSize, fb.s, fb.e, fb.p);
          }
        } catch {
          // ignore malformed feedback
        }
      }
    });

    readStream.pipe(socket, { end: false });

    readStream.on('end', () => {
      if (cancelled) {
        this.activeSends.delete(transferId);
        return;
      }
      streamDone = true;
      socket.end();
      socket.once('close', (hadError: boolean) => {
        if (!cancelled && !hadError) {
          this.emit('progress', transferId, fileSize, fileSize, 0, 0, 100);
          this.emit('complete', transferId);
        }
        this.activeSends.delete(transferId);
      });
    });

    readStream.on('error', (err) => {
      if (!cancelled) {
        socket.destroy(err);
        this.emit('error', transferId, err);
      }
      this.activeSends.delete(transferId);
    });

    void streamDone;
  }

  /** Encrypted streaming — reads file in TRANSFER_CHUNK_SIZE chunks, encrypts each. */
  private streamFileEncrypted(
    socket: Socket,
    transferId: string,
    filePath: string,
    fileSize: number,
    resumeOffset = 0,
    encryptionKey: Buffer,
  ): void {
    const CHUNK_SIZE = 1024 * 1024; // 1 MB — matches TRANSFER_CHUNK_SIZE
    let cancelled = false;

    const existing = this.activeSends.get(transferId);
    if (existing) {
      this.activeSends.set(transferId, {
        ...existing,
        cancel: () => {
          cancelled = true;
          socket.destroy();
        },
      });
    }

    // Feedback listener (same as plaintext path)
    let feedbackBuffer = '';
    socket.on('data', (chunk: Buffer) => {
      feedbackBuffer += chunk.toString('utf8');
      let newlineIdx: number;
      while ((newlineIdx = feedbackBuffer.indexOf('\n')) !== -1) {
        const line = feedbackBuffer.slice(0, newlineIdx);
        feedbackBuffer = feedbackBuffer.slice(newlineIdx + 1);
        try {
          const fb = JSON.parse(line) as { p: number; b: number; s: number; e: number };
          if (typeof fb.p === 'number') {
            this.emit('progress', transferId, fb.b, fileSize, fb.s, fb.e, fb.p);
          }
        } catch {
          // ignore malformed feedback
        }
      }
    });

    // Read and encrypt the file asynchronously
    const doEncryptedStream = async (): Promise<void> => {
      const { createReadStream } = await import('fs');
      const readStream = createReadStream(filePath, {
        start: resumeOffset > 0 ? resumeOffset : 0,
        highWaterMark: CHUNK_SIZE,
      });

      for await (const rawChunk of readStream) {
        if (cancelled) return;
        const chunk = Buffer.isBuffer(rawChunk) ? rawChunk : Buffer.from(rawChunk as Uint8Array);
        const encrypted = TransferCrypto.encryptChunk(encryptionKey, chunk);
        const encoded = TransferCrypto.encodeChunk(encrypted);

        // Respect backpressure
        const canContinue = socket.write(encoded);
        if (!canContinue) {
          await new Promise<void>((resolve) => socket.once('drain', resolve));
        }
      }

      if (!cancelled) {
        socket.end();
        await new Promise<void>((resolve) => socket.once('close', resolve));
        if (!cancelled) {
          this.emit('progress', transferId, fileSize, fileSize, 0, 0, 100);
          this.emit('complete', transferId);
        }
      }
    };

    doEncryptedStream().catch((err: Error) => {
      if (!cancelled) {
        socket.destroy(err);
        this.emit('error', transferId, err);
      }
      this.activeSends.delete(transferId);
    });
  }
}

File: AndroidApp/android/app/src/main/java/com/wyre/app/TransferClient.kt
package com.wyre.app

import android.util.Base64
import java.io.File
import java.net.Socket
import java.security.MessageDigest

private const val CHUNK_SIZE = 1 * 1024 * 1024
private const val PROGRESS_INTERVAL_MS = 150L

/**
 * TransferClient.kt
 * TCP client — connects to a peer's TransferServer and streams a file.
 *
 * When peerSupportsEncryption is true, performs an ECDH P-256 handshake and
 * encrypts each chunk with AES-256-GCM before writing to the socket.
 * Falls back to plaintext if the peer does not accept encryption.
 */
class TransferClient(
    private val transferId: String,
    private val filePath: String,
    private val fileName: String,
    private val fileSize: Long,
    private val peerIp: String,
    private val peerPort: Int,
    private val senderDeviceId: String,
    private val senderName: String,
    private val resumeOffset: Long = 0,
    private val peerSupportsEncryption: Boolean = false,
    private val onEvent: (TransferEvent) -> Unit
) : Cancellable {

    @Volatile private var cancelled = false
    private var socket: Socket? = null

    override fun cancel() {
        cancelled = true
        socket?.close()
    }

    fun send() {
        val startedAt = System.currentTimeMillis()

        val file = File(filePath)
        val actualFileSize = if (fileSize > 0) fileSize else file.length()

        onEvent(TransferEvent.Started(
            transferId = transferId,
            direction  = "send",
            peerId     = "",
            peerName   = senderName,
            fileName   = fileName,
            fileSize   = actualFileSize,
            status     = "connecting"
        ))

        try {
            val sock = Socket(peerIp, peerPort)
            socket = sock

            val out = sock.getOutputStream()
            val inp = sock.getInputStream()

            // ── 1. Build and send JSON header ─────────────────────────────────
            var senderPubKeyDer: ByteArray? = null
            var senderPrivKey: java.security.PrivateKey? = null

            val headerJson = org.json.JSONObject().apply {
                put("transferId",      transferId)
                put("senderDeviceId",  senderDeviceId)
                put("senderName",      senderName)
                put("fileName",        fileName)
                put("fileSize",        actualFileSize)
                put("checksum",        computeChecksum())
                if (resumeOffset > 0) put("resumeOffset", resumeOffset)

                if (peerSupportsEncryption) {
                    val (pubKeyDer, privKey) = TransferCrypto.generateKeyPair()
                    senderPubKeyDer = pubKeyDer
                    senderPrivKey   = privKey
                    put("encryption", org.json.JSONObject().apply {
                        put("supported",       true)
                        put("senderPublicKey", Base64.encodeToString(pubKeyDer, Base64.NO_WRAP))
                    })
                }
            }

            out.write((headerJson.toString() + "\n").toByteArray(Charsets.UTF_8))
            out.flush()

            // ── 2. Read accept/decline response ───────────────────────────────
            val responseLine = readLine(inp) ?: throw Exception("No response from peer")
            val responseJson = org.json.JSONObject(responseLine)
            val accepted = responseJson.optBoolean("accepted", false)
            if (!accepted) {
                onEvent(TransferEvent.Error(transferId, "Declined by recipient", "DECLINED"))
                sock.close()
                return
            }

            // ── 3. Negotiate encryption ───────────────────────────────────────
            var encryptionKey: ByteArray? = null

            if (peerSupportsEncryption && senderPubKeyDer != null && senderPrivKey != null) {
                val encResp = responseJson.optJSONObject("encryption")
                if (encResp != null && encResp.optBoolean("accepted", false)) {
                    val receiverPubKeyB64 = encResp.optString("receiverPublicKey", "")
                    if (receiverPubKeyB64.isNotEmpty()) {
                        try {
                            val receiverPubKeyDer = Base64.decode(receiverPubKeyB64, Base64.NO_WRAP)
                            encryptionKey = TransferCrypto.deriveKey(
                                privateKey         = senderPrivKey!!,
                                peerPublicKeyDer   = receiverPubKeyDer,
                                senderPublicKeyDer = senderPubKeyDer!!,
                                receiverPublicKeyDer = receiverPubKeyDer,
                            )
                        } catch (_: Exception) {
                            // Key derivation failed — fall back to plaintext
                            encryptionKey = null
                        }
                    }
                }
            }

            onEvent(TransferEvent.Started(
                transferId = transferId,
                direction  = "send",
                peerId     = "",
                peerName   = senderName,
                fileName   = fileName,
                fileSize   = actualFileSize,
                status     = "active"
            ))

            // ── 4. Stream file ────────────────────────────────────────────────
            if (encryptionKey != null) {
                streamEncrypted(sock, out, inp, file, actualFileSize, encryptionKey!!, startedAt)
            } else {
                streamPlaintext(sock, out, inp, file, actualFileSize, startedAt)
            }

        } catch (e: Exception) {
            if (!cancelled) {
                onEvent(TransferEvent.Error(transferId, e.message ?: "Send failed", "SEND_ERROR"))
            } else {
                onEvent(TransferEvent.Error(transferId, "Cancelled", "CANCELLED"))
            }
        }
    }

    // ── Plaintext streaming (original behaviour) ──────────────────────────────

    private fun streamPlaintext(
        sock: Socket,
        out: java.io.OutputStream,
        inp: java.io.InputStream,
        file: File,
        actualFileSize: Long,
        startedAt: Long,
    ) {
        var bytesSent = resumeOffset
        var lastProgressTime = System.currentTimeMillis()
        var lastProgressBytes = resumeOffset
        val buf = ByteArray(CHUNK_SIZE)
        var chunksSinceFlush = 0

        val fis = java.io.FileInputStream(file)
        if (resumeOffset > 0) {
            var skipped = 0L
            while (skipped < resumeOffset) {
                val n = fis.skip(resumeOffset - skipped)
                if (n <= 0) break
                skipped += n
            }
        }
        fis.use { fis ->
            var read: Int
            while (fis.read(buf).also { read = it } != -1) {
                if (cancelled) {
                    sock.close()
                    onEvent(TransferEvent.Error(transferId, "Cancelled", "CANCELLED"))
                    return
                }

                out.write(buf, 0, read)
                bytesSent += read
                chunksSinceFlush++

                if (chunksSinceFlush >= 8) {
                    out.flush()
                    chunksSinceFlush = 0
                }

                emitProgressIfDue(bytesSent, actualFileSize, lastProgressTime, lastProgressBytes) { time, bytes ->
                    lastProgressTime = time
                    lastProgressBytes = bytes
                }
            }
        }

        out.flush()
        sock.shutdownOutput()
        drainFeedback(inp)

        if (!cancelled) {
            onEvent(TransferEvent.Progress(transferId, 100, 0, 0, actualFileSize, actualFileSize))
            onEvent(TransferEvent.Complete(
                transferId = transferId,
                direction  = "send",
                peerId     = "",
                peerName   = senderName,
                fileName   = fileName,
                fileSize   = actualFileSize,
                savedPath  = "",
                startedAt  = startedAt
            ))
        }
        sock.close()
    }

    // ── Encrypted streaming ───────────────────────────────────────────────────

    private fun streamEncrypted(
        sock: Socket,
        out: java.io.OutputStream,
        inp: java.io.InputStream,
        file: File,
        actualFileSize: Long,
        encryptionKey: ByteArray,
        startedAt: Long,
    ) {
        var bytesSent = resumeOffset
        var lastProgressTime = System.currentTimeMillis()
        var lastProgressBytes = resumeOffset
        val buf = ByteArray(CHUNK_SIZE)

        val fis = java.io.FileInputStream(file)
        if (resumeOffset > 0) {
            var skipped = 0L
            while (skipped < resumeOffset) {
                val n = fis.skip(resumeOffset - skipped)
                if (n <= 0) break
                skipped += n
            }
        }
        fis.use { fis ->
            var read: Int
            while (fis.read(buf).also { read = it } != -1) {
                if (cancelled) {
                    sock.close()
                    onEvent(TransferEvent.Error(transferId, "Cancelled", "CANCELLED"))
                    return
                }

                val plaintext = buf.copyOf(read)
                val (iv, ciphertext, tag) = TransferCrypto.encryptChunk(encryptionKey, plaintext)
                val encoded = TransferCrypto.encodeChunk(iv, ciphertext, tag)
                out.write(encoded)
                out.flush()

                bytesSent += read

                emitProgressIfDue(bytesSent, actualFileSize, lastProgressTime, lastProgressBytes) { time, bytes ->
                    lastProgressTime = time
                    lastProgressBytes = bytes
                }
            }
        }

        out.flush()
        sock.shutdownOutput()
        drainFeedback(inp)

        if (!cancelled) {
            onEvent(TransferEvent.Progress(transferId, 100, 0, 0, actualFileSize, actualFileSize))
            onEvent(TransferEvent.Complete(
                transferId = transferId,
                direction  = "send",
                peerId     = "",
                peerName   = senderName,
                fileName   = fileName,
                fileSize   = actualFileSize,
                savedPath  = "",
                startedAt  = startedAt
            ))
        }
        sock.close()
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private inline fun emitProgressIfDue(
        bytesSent: Long,
        actualFileSize: Long,
        lastProgressTime: Long,
        lastProgressBytes: Long,
        update: (Long, Long) -> Unit,
    ) {
        val now = System.currentTimeMillis()
        val elapsed = now - lastProgressTime
        if (bytesSent > 0 && elapsed >= PROGRESS_INTERVAL_MS) {
            val bytesInInterval = bytesSent - lastProgressBytes
            val speed = if (elapsed > 0) (bytesInInterval * 1000L / elapsed) else 0L
            val progress = if (actualFileSize > 0) ((bytesSent * 100L) / actualFileSize).toInt().coerceIn(1, 99) else 1
            val eta = if (speed > 0) (actualFileSize - bytesSent) / speed else 0L
            onEvent(TransferEvent.Progress(
                transferId       = transferId,
                progress         = progress,
                speed            = speed,
                eta              = eta,
                bytesTransferred = bytesSent,
                totalBytes       = actualFileSize
            ))
            update(now, bytesSent)
        }
    }

    private fun drainFeedback(inp: java.io.InputStream) {
        try {
            val drainBuf = ByteArray(4096)
            while (inp.read(drainBuf) != -1) { /* drain */ }
        } catch (_: Exception) {}
    }

    private fun buildHeader(actualFileSize: Long): String = org.json.JSONObject().apply {
        put("transferId",      transferId)
        put("senderDeviceId",  senderDeviceId)
        put("senderName",      senderName)
        put("fileName",        fileName)
        put("fileSize",        actualFileSize)
        put("checksum",        computeChecksum())
        if (resumeOffset > 0) put("resumeOffset", resumeOffset)
    }.toString()

    private fun computeChecksum(): String {
        val md = MessageDigest.getInstance("SHA-256")
        File(filePath).inputStream().use { fis ->
            val buf = ByteArray(CHUNK_SIZE)
            var read: Int
            while (fis.read(buf).also { read = it } != -1) md.update(buf, 0, read)
        }
        return md.digest().joinToString("") { "%02x".format(it) }
    }

    private fun readLine(inp: java.io.InputStream): String? {
        val sb = StringBuilder()
        var b: Int
        while (inp.read().also { b = it } != -1) {
            if (b == '\n'.code) return sb.toString()
            sb.append(b.toChar())
        }
        return if (sb.isNotEmpty()) sb.toString() else null
    }
}
